python/dgl/nn/pytorch/softmax.py

```diff
@@ -50,46 +47,86 @@ def forward(self, logits, graph):
 
         Notes
         -----
-            * Input shape: :math:`(N, *, 1)` where * means any number of additional
-              dimensions, :math:`N` is the number of edges.
-            * Unnormalized scores shape: :math:`(N, *, 1)` where all but the last
-              dimension are the same shape as the input.
-            * Normalizer shape: :math:`(M, *, 1)` where :math:`M` is the number of
-              nodes and all but the first and the last dimensions are the same as
-              the input.
+            * Input shape: :math:`(N, *, 1)` where * means any number of
+              additional dimensions, :math:`N` is the number of edges.
+            * Unnormalized scores shape: :math:`(N, *, 1)` where all but the
+              last dimension are the same shape as the input.
+            * Normalizer shape: :math:`(M, *, 1)` where :math:`M` is the number
+              of nodes and all but the first and the last dimensions are the
+              same as the input.
 
-        Note that this computation is still one step away from getting real softmax
-        results. The last step can be proceeded as follows:
+        Note that this computation is still one step away from getting real
+        softmax results. The last step can be proceeded as follows:
 
         >>> import dgl.function as fn
-        >>>
-        >>> scores, normalizer = EdgeSoftmax(...).forward(logits, graph)
+        >>> scores, normalizer = EdgeSoftmax(logits, graph)
         >>> graph.edata['a'] = scores
         >>> graph.ndata['normalizer'] = normalizer
-        >>> graph.apply_edges(lambda edges : {'a' : edges.data['a'] / edges.dst['normalizer']})
+        >>> graph.apply_edges(
+                lambda edges: {'a': edges.data['a'] / edges.dst['normalizer']})
 
-        We left this last step to users as depending on the particular use case,
-        this step can be combined with other computation at once.
+        We left this last step to users as depending on the particular use
+        case, this step can be combined with other computation at once.
+        """
+        num_nodes = graph.number_of_nodes()
+        ctx = utils.to_dgl_context(F.context(logits))
+        gidx = graph._graph.get_immutable_gidx(ctx)
+        _, dst, _ = graph._graph.edges()
+        dst = dst.tousertensor(F.context(logits))
+        empty_map = (None, None)
+        max_logits_ = F.copy_reduce("max", gidx, fn.TargetCode.EDGE, logits,
+                                    num_nodes, empty_map, empty_map)
+        logits = (logits - max_logits_.index_select(0, dst)).exp()
+        norm = F.copy_reduce("sum", gidx, fn.TargetCode.EDGE, logits,
+                             num_nodes, empty_map, empty_map)
+        return logits / norm.index_select(0, dst)
```

Comment:
I assume that this class should be removed and replaced with the class below.

Reply:
I suppose so. Will remove and we benchmark the speed.

Reply:
Do we still need to write the backward pass for edge softmax?

Reply:
Yes, so we could avoid backward the max operation for example.